### PR TITLE
feat: add speculator training components

### PR DESCRIPTION
**Pull request by hrathina**
## Summary

Adds the Speculator KFP component family wrapping Kubeflow Trainer's `SpeculativeDecodingTrainer`. The component runtime supports the four SDK modes: `data_only`, `train_only`, `offline`, and `online`.

## Changes

- Added Speculator component ownership and package structure under `components/training/speculator/`.
- Added dedicated extraction, training, and online component entry points.
- Added Speculator-owned shared runtime helpers for setup, TrainJob handling, model output persistence, and mode dispatch.
- Added the model download toggle needed by Speculator model caching.
- Preserved fine-tuning component ownership and shared utilities.

## Validation

- Pre-commit hooks passed.
- Component code was linted with Ruff.
- Speculator pipeline compilation and mode-specific tests were run during development.

## Scope

This PR contains component changes only. Pipeline and unrelated configuration changes remain outside this commit.

[RHOAIENG-88185]: https://redhat.atlassian.net/browse/RHOAIENG-88185?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

- **New Features**
  - Added Speculator workflows for extracting hidden states, training draft models, and combined online extraction and training.
  - Added support for managed or external vLLM services, configurable training, metrics, storage, and resource options.
  - Added an option to skip model downloads while returning the expected model location.

- **Documentation**
  - Added guides and metadata for Speculator workflows and linked them from training documentation.
  - Updated supported base-image configuration.

- **Tests**
  - Added coverage for Speculator workflows and optional model-download behavior.

### CODE REVIEW

File: components/training/speculator/shared/output.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+def persist_model(ckpt_dir: str, pvc_path: str, model_name: str, output_model, log: logging.Logger) -> None:
+    """Copy the trained model to the PVC and KFP output artifact."""
+    latest = find_model_dir(ckpt_dir)
+    if not latest:
+        raise RuntimeError(f"No model found in {ckpt_dir}")
+    pvc_output = os.path.join(pvc_path, "final_model")
+    if os.path.exists(pvc_output):
+        shutil.rmtree(pvc_output, ignore_errors=True)
+    shutil.copytree(latest, pvc_output, dirs_exist_ok=True)
```
[the remark this answers is not in the document]
Comment (hrathina) [reply components/training/speculator/shared/output.py]: This finding is no longer applicable. The component now selects only `<output_dir>/checkpoint_best`; it cannot select `<pvc_path>/final_model` as the source. If `checkpoint_best` is missing, the component fails instead of selecting another directory, so it cannot delete its own source before copying.

File: components/training/speculator/shared/output.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    shutil.copytree(latest, output_model.path, dirs_exist_ok=True)
```
[the remark this answers is not in the document]
Comment (hrathina) [reply components/training/speculator/shared/output.py]: Valid finding. If removing `final_model` fails, `dirs_exist_ok=True` can merge stale files into the published model directory. The cleanup should fail on removal errors and copy into a clean destination without merging.

File: components/training/finetuning/speculator/extraction/tests/__init__.py
Comment (efazal) [inline components/training/finetuning/speculator/extraction/tests/__init__.py:1]: The `speculator/` folder should be created under the finetuning `components/training/finetuning/`

File: components/training/finetuning/speculator/extraction/OWNERS
Comment (efazal) [inline components/training/finetuning/speculator/extraction/OWNERS:1]: The owners file should follow this pattern seen here `components/training/finetuning/OWNERS` and the names should be in alphabetical order.

File: components/training/finetuning/speculator/training/README.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+| `persistent_mount_path` | `str` | `/mnt/persistent` |  |
+| `output_model` | `dsl.Output[dsl.Model]` | `None` |  |
+| `output_metrics` | `dsl.Output[dsl.Metrics]` | `None` |  |
+| `kubernetes_config` | `dsl.TaskConfig` | `None` |  |
+
+## Outputs 📤
+
+| Name | Type | Description |
+| ---- | ---- | ----------- |
+| Output | `str` |  |
```
[the remark this answers is not in the document]
Comment (hrathina) [reply components/training/finetuning/speculator/training/README.md:61]: This is generated from the component's → str return annotation and matches the existing SFT/LoRA README pattern. output_model and output_metrics
are already documented as component parameters. Manually changing the Outputs table would make the README fail the repository synchronization
check, so no change is needed.

File: components/training/speculator/shared/setup.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    if not server or not token:
+        raise RuntimeError("KUBERNETES_SERVER_URL and KUBERNETES_AUTH_TOKEN must be set")
+    ca_path = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+    if not os.path.isfile(ca_path):
+        raise RuntimeError(f"In-cluster CA certificate not found at {ca_path}")
+    config = k8s.Configuration()
+    config.host = server
+    config.verify_ssl = True
+    config.ssl_ca_cert = ca_path
+    config.api_key = {"authorization": f"Bearer {token}"}
```
[the remark this answers is not in the document]
Comment (hrathina) [reply components/training/speculator/shared/setup.py]: This Kubernetes setup follows the existing SFT, LoRA, and OSFT component

File: components/training/speculator/shared/output.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+
+def find_model_dir(root: str):
+    """Find the best model checkpoint produced by Speculator training."""
+    if not os.path.isdir(root):
+        return None
+    best_dir = os.path.join(root, "checkpoint_best")
+    if os.path.isfile(os.path.join(best_dir, "config.json")):
+        return best_dir
+    return None
```
[the remark this answers is not in the document]
Comment (hrathina) [reply components/training/speculator/shared/output.py]: Verified against the linked Speculators Trainer implementation. maybe_update_best() calls checkpointer.update_best_symlink(epoch) outside the save_best conditional. save_best only controls best-checkpoint retention and cleanup. Since TRAIN_ONLY and ONLINE provide validation loaders, checkpoint_best is created when validation improves, so requiring checkpoint_best is correct.

File: components/training/finetuning/speculator/extraction/tests/test_component_unit.py
Comment (efazal) [inline components/training/finetuning/speculator/extraction/tests/test_component_unit.py:1]: The shared folder also should be within the `speculator/` module

Comment (hrathina) [reply components/training/finetuning/speculator/extraction/tests/test_component_unit.py:1]: The shared folder is common for all fine-tuning pipeline. I will move the speculator file inside the speculator/ module.

Comment (hrathina) [reply components/training/finetuning/speculator/extraction/tests/test_component_unit.py:1]: Changes done.

**Comment by efazal** [comment]
Thanks @hrathina — the component structure is clean. A few things to address before merge, mostly around the shared package duplication and some silent failure paths in the speculator dispatcher.

File: components/training/finetuning/speculator/shared/setup.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
+"""Setup utilities: logging, K8s initialization, environment config, HF token."""
```
Comment (efazal) [inline components/training/finetuning/speculator/shared/setup.py:1]: `setup.py` and `training.py` here are byte-identical to the ones in `components/training/finetuning/shared/`, which already serves sft/osft/lora. The only genuine addition is the `prefer_best`/`strict_output` args in `output.py`, which are backward-compatible and could go directly into the existing `finetuning/shared/output.py`.

This means any future bug fix to `finetuning/shared/training.py` needs to be applied to this copy too, and this copy has 3 string-matching tests vs 107 behavioral tests in the original. Consider upstreaming the `output.py` additions and deleting this directory, pointing `_SHARED_DIR` at `"../../shared"` instead.

File: components/training/finetuning/speculator/shared/speculator.py [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+    if mode == "data_only":
+        hidden_states_dir = local_path(
+            values["hidden_states_path"] if external else values["output_dir"] + "/hidden_states"
+        )
+        output = values.get("output_hidden_states")
+        if output:
+            output.uri = hidden_states_dir
+            output.metadata["pvc_path"] = hidden_states_dir
+        return "data_only completed - hidden states saved"
+    model_dir = os.path.join(local_path(values["output_dir"]), "checkpoint_best")
```
Comment (efazal) [inline components/training/finetuning/speculator/shared/speculator.py]: `model_dir` already ends in `checkpoint_best`, so inside `persist_model`, `prefer_best=True` looks for `checkpoint_best/checkpoint_best` — a path that never exists — and silently falls through to `find_model_dir()`. That heuristic picks the most recently modified `config.json` in the tree, which may be a tokenizer or verifier subdirectory rather than the model root.

Either pass `local_path(values["output_dir"])` and let `persist_model` append `checkpoint_best`, or drop `prefer_best=True` from the call.

File: components/training/finetuning/speculator/shared/speculator.py [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+            client,
+            job,
+            log,
+        )
+    except Exception:
+        log.error("Speculator %s failed", mode)
+        raise
+
+    if mode == "data_only":
+        hidden_states_dir = local_path(
```
Comment (efazal) [inline components/training/finetuning/speculator/shared/speculator.py:390]: The hidden states directory is assumed to be `output_dir/hidden_states` with no existence check. If the Speculator SDK writes to a different path, `output.uri` is set to a nonexistent directory and the component returns `"data_only completed — hidden states saved"` as a success. The downstream `train_only` step then fails with a missing-path error with no pointer back to this step.

Validate `os.path.isdir(hidden_states_dir)` before setting the artifact URI, and raise with a message pointing to the job logs.

File: components/training/finetuning/speculator/shared/speculator.py [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+            values["hidden_states_path"] if external else values["output_dir"] + "/hidden_states"
+        )
+        output = values.get("output_hidden_states")
+        if output:
+            output.uri = hidden_states_dir
+            output.metadata["pvc_path"] = hidden_states_dir
```
Comment (efazal) [inline components/training/finetuning/speculator/shared/speculator.py:401]: When `output_hidden_states` is not wired in the pipeline, the component silently returns `"data_only completed — hidden states saved"` with no artifact set and no log warning. The return string is actively misleading. Add a `log.warning` naming the PVC path so the operator knows the output is PVC-only and not consumable by downstream KFP steps.

File: components/training/finetuning/speculator/shared/speculator.py [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+        "training_scheduler_total_steps",
+    ):
+        if values.get(key) == 0:
+            values[key] = None
+
+    if mode == "data_only" and values.get("vllm_source") == "remote":
+        values["verifier_model"] = values.get("verifier_model_pvc") or values["verifier_model"]
+    elif mode in ("train_only", "online") and values.get("verifier_model_pvc"):
+        values["verifier_model"] = values["verifier_model_pvc"]
+    external = values.get("vllm_endpoint") is not None
```
Comment (efazal) [inline components/training/finetuning/speculator/shared/speculator.py:133]: The cleanup loop above coerces any falsy `vllm_endpoint` value (including `""`) to `None`, so a user who passes an empty string silently switches from external to internal vLLM mode and the validation block at lines 145-148 is never reached.

Explicitly reject empty string before the coercion loop with a clear message.

File: components/training/finetuning/speculator/shared/speculator.py [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+        or values["verifier_model"].startswith("pvc://")
+        else values["verifier_model"]
+    )
+    if values.get("download_model") and not external:
+        model_dir_name = values["verifier_model"].replace("/", "--")
+        trainer_verifier_model = trainer_path(f"models/{model_dir_name}")
+    trainer_output_dir = trainer_path(values["output_dir"])
+    trainer_hidden_states_path = trainer_path(values.get("hidden_states_path"))
+    trainer_training_data_path = trainer_path(values.get("training_data_path"))
+    if external:
```
Comment (efazal) [inline components/training/finetuning/speculator/shared/speculator.py:149]: `trainer_verifier_model` is always produced by `trainer_path()`, which unconditionally rewrites any non-`pvc://` URI to `pvc://...`. So `trainer_verifier_model.startswith("pvc://")` can never be false — this half of the guard is vacuous. A user passing a HuggingFace model ID as `verifier_model` with an external endpoint gets a wrong `pvc://workspace/...` path forwarded to the Trainer SDK instead of a clear error.

Validate `values["verifier_model"]` (the raw input) directly.

File: components/training/finetuning/speculator/shared/speculator.py [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+            mounts = [api_client.sanitize_for_serialization(x) if not isinstance(x, dict) else x for x in mounts]
+            # The Trainer SDK mounts every PVC URI at its checkpoint path. Do not
+            # add the same user PVC again at the launcher mount path when the
+            # output/model storage uses that PVC.
+            storage_claims = set()
+            for storage_uri in (trainer_output_dir, trainer_hidden_states_path, trainer_training_data_path):
+                claim, _ = pvc_parts(storage_uri or "")
+                if claim:
+                    storage_claims.add(claim)
+            duplicate_volume_names = {
```
Comment (efazal) [inline components/training/finetuning/speculator/shared/speculator.py]: Deduplication matches on `claimName` alone. If the same PVC is mounted twice for different purposes (e.g., workspace at `/mnt/ws` and a dataset subPath at `/mnt/data`), both mounts are stripped. The TrainJob then fails on a missing path with no connection to this filter. Match on claim + mount path pair and log which volumes were removed.

File: components/training/finetuning/speculator/shared/speculator.py [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+    for key in (
+        "data_extraction_max_samples",
+        "training_draft_vocab_size",
+        "training_scheduler_warmup_steps",
+        "training_scheduler_total_steps",
+    ):
+        if values.get(key) == 0:
+            values[key] = None
+
+    if mode == "data_only" and values.get("vllm_source") == "remote":
```
Comment (efazal) [inline components/training/finetuning/speculator/shared/speculator.py:129]: `vllm_source` is never set by any component — confirmed grep across the whole repo — so this branch is permanently dead. `mode == "offline"` at line 132 likewise never fires since the three call sites hardcode `"data_only"`, `"train_only"`, and `"online"`.

Delete both dead branches. If it is worth documenting that OFFLINE is a pipeline-level composition of `data_only` + `train_only`, a one-line comment is clearer than dead executable code.

File: components/training/finetuning/speculator/shared/speculator.py [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+        else:
+            options = None
+        job = client.train(trainer=trainer, options=options, runtime=runtime)
+        wait_for_training_job(
+            client,
+            job,
+            log,
+        )
+    except Exception:
+        log.error("Speculator %s failed", mode)
```
Comment (efazal) [inline components/training/finetuning/speculator/shared/speculator.py]: nit: `log.error` drops the traceback. Use `log.exception(...)` so the exception type and stack are visible in pipeline logs.

File: components/training/finetuning/speculator/shared/speculator.py [hunk 15]
```diff
@@ -0,0 +1,391 @@
...
+            model_dir,
+            persistent_mount_path if persistent_pvc else pvc_path,
+            values["verifier_model"],
+            values["output_model"],
+            log,
+            prefer_best=True,
+            strict_output=True,
+        )
+    if values.get("output_metrics"):
+        values["output_metrics"].log_metric("mode", mode)
```
Comment (efazal) [inline components/training/finetuning/speculator/shared/speculator.py]: nit: `log_metric` expects a float; passing a string (`mode`) will fail or produce an invalid Metrics artifact depending on the KFP runtime version. Either omit this metric or encode mode as a float (e.g., an index).

File: components/training/finetuning/speculator/online/component.py [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    vllm_resource_memory: str = "96Gi",
+    vllm_gpu_memory_utilization: float = 0.9,
+    vllm_readiness_timeout_minutes: int = 60,
+    enable_progression_tracking: bool = True,
+    metrics_port: int = 28080,
+    metrics_poll_interval_seconds: int = 30,
+    packages_to_install: Optional[str] = None,
+    pip_index_urls: Optional[str] = None,
+    training_envs: str = "",
+    training_runtime: str = "vllm-extract-cuda",
```
Comment (efazal) [inline components/training/finetuning/speculator/online/component.py:73]: `"vllm-extract-cuda"` does not appear anywhere else in the repo — no ClusterTrainingRuntime definition, no reference. Every sibling component uses `"speculator-model-opt-cuda"`. On default invocation `select_runtime` will raise `Runtime 'vllm-extract-cuda' not found`. Please confirm the correct runtime name.

File: components/training/finetuning/speculator/shared/setup.py (hunk 6 above)
Comment (hrathina) [reply components/training/finetuning/speculator/shared/setup.py:1]: Thanks for pointing this out. I agree that duplicating the shared helpers is not ideal. I tested embedding `finetuning/shared/` directly, but it increased the KFP/Argo payload enough to risk the `Argument list too long failure.` The smaller` speculator/shared/` bundle keeps the training component payload within limits and is required for the current lightweight component packaging. I kept the bundle intentionally minimal and aligned its helper implementations with` finetuning/ shared`/. A follow-up cleanup could move this runtime into a dedicated container image, which would remove the duplication without reintroducing the payload-size issue.

File: components/training/finetuning/speculator/online/component.py (hunk 16 above)
Comment (hrathina) [reply components/training/finetuning/speculator/online/component.py:73]: Confirmed. vllm-extract-cuda is the correct ClusterTrainingRuntime name for managed VLLM extraction.
No code change is needed.
